optimize: size produces invalid typescript

I believe  this was the cause of #12, but wanted to open another issue as that one is closed and almost a year old.

If I clone this repo and include `--optimize size` in the `test:cli` script, I get:
```
> pegjs --plugin ./src/tspegjs --extra-options-file test/genoptions2.json  --optimize size --allowed-start-rules groupFile,templateFile,templateFileRaw,templateAndEOF -o output/st2.ts examples/st.pegjs && tsc output/st2.ts && tslint output/st2.ts

output/st2.ts(938,5): error TS2540: Cannot assign to 'peg$startRuleIndex' because it is a constant or a read-only property.
output/st2.ts(1067,5): error TS2322: Type 'number[]' is not assignable to type 'number'.
output/st2.ts(1074,18): error TS2339: Property 'length' does not exist on type 'number'.
output/st2.ts(1220,66): error TS2339: Property 'length' does not exist on type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...'.
  Property 'length' does not exist on type 'RegExp'.
output/st2.ts(1234,66): error TS2339: Property 'length' does not exist on type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...'.
  Property 'length' does not exist on type 'RegExp'.
output/st2.ts(1248,40): error TS2339: Property 'test' does not exist on type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...'.
  Property 'test' does not exist on type 'string'.
output/st2.ts(1266,51): error TS2339: Property 'length' does not exist on type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...'.
  Property 'length' does not exist on type 'RegExp'.
output/st2.ts(1273,24): error TS2345: Argument of type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...' is not assignable to parameter of type 'Expectation'.
  Type 'string' is not assignable to type 'Expectation'.
output/st2.ts(1289,25): error TS2339: Property 'slice' does not exist on type 'number'.
output/st2.ts(1295,38): error TS2339: Property 'apply' does not exist on type 'string | RegExp | ILiteralExpectation | IClassExpectation | IAnyExpectation | IOtherExpectation |...'.
  Property 'apply' does not exist on type 'string'.
```

I am currently converting my library to typescript, and this library has been tremendously helpful, but I had to remove `optimize: size`, which I believe is the main contributor to the doubling in size of the bundle  unminified, and 25% increase minified (its about an order of magnitude larger) minified. Bundles aren't necessarily the main way the library will be used going forward, but it's hard to fight for a size increase.

--- a/src/passes/generate-ts.js
+++ b/src/passes/generate-ts.js
@@ -190,7 +190,7 @@ function generateTS(ast, options) {
         "stack.splice(",
         "  stack.length - bc[ip + 2],",
         "  bc[ip + 2],",
-        "  peg$consts[bc[ip + 1]].apply(null, params)",
+        "  (peg$consts[bc[ip + 1]] as ((...args) => any)).apply(null, params)",
         ");",
         "",
         "ip += " + baseLength + " + " + paramsLengthCode + ";",
@@ -199,8 +199,8 @@ function generateTS(ast, options) {
     }
 
     parts.push([
-      "function peg$decode(s: string): number {",
-      "  return s.split(\"\").map((ch) => { return ch.charCodeAt(0) - 32; });",
+      "function peg$decode(s: string): number[] {",
+      "  return s.split(\"\").map((ch) =>  ch.charCodeAt(0) - 32 );",
       "}",
       "",
       "function peg$parseRule(index) {"
@@ -208,23 +208,23 @@ function generateTS(ast, options) {
 
     if (options.trace) {
       parts.push([
-        "  let bc = peg$bytecode[index];",
+        "  const bc = peg$bytecode[index];",
         "  let ip = 0;",
-        "  let ips = [];",
+        "  const ips = [];",
         "  let end = bc.length;",
-        "  let ends = [];",
-        "  let stack = [];",
+        "  const ends = [];",
+        "  const stack = [];",
         "  let startPos = peg$currPos;",
         "  let params;"
       ].join("\n"));
     } else {
       parts.push([
-        "  let bc = peg$bytecode[index];",
+        "  const bc = peg$bytecode[index];",
         "  let ip = 0;",
-        "  let ips = [];",
+        "  const ips = [];",
         "  let end = bc.length;",
-        "  let ends = [];",
-        "  let stack = [];",
+        "  const ends = [];",
+        "  const stack = [];",
         "  let params;"
       ].join("\n"));
     }
@@ -328,19 +328,19 @@ function generateTS(ast, options) {
       "",
       "        case " + op.MATCH_STRING + ":", // MATCH_STRING s, a, f, ...
       indent10(generateCondition(
-        "input.substr(peg$currPos, peg$consts[bc[ip + 1]].length) === peg$consts[bc[ip + 1]]",
+        "input.substr(peg$currPos, (peg$consts[bc[ip + 1]] as string).length) === peg$consts[bc[ip + 1]]",
         1
       )),
       "",
       "        case " + op.MATCH_STRING_IC + ":", // MATCH_STRING_IC s, a, f, ...
       indent10(generateCondition(
-        "input.substr(peg$currPos, peg$consts[bc[ip + 1]].length).toLowerCase() === peg$consts[bc[ip + 1]]",
+        "input.substr(peg$currPos, (peg$consts[bc[ip + 1]] as string).length).toLowerCase() === peg$consts[bc[ip + 1]]",
         1
       )),
       "",
       "        case " + op.MATCH_REGEXP + ":", // MATCH_REGEXP r, a, f, ...
       indent10(generateCondition(
-        "peg$consts[bc[ip + 1]].test(input.charAt(peg$currPos))",
+        "(peg$consts[bc[ip + 1]] as RegExp).test(input.charAt(peg$currPos))",
         1
       )),
       "",
@@ -352,14 +352,14 @@ function generateTS(ast, options) {
       "",
       "        case " + op.ACCEPT_STRING + ":", // ACCEPT_STRING s
       "          stack.push(peg$consts[bc[ip + 1]]);",
-      "          peg$currPos += peg$consts[bc[ip + 1]].length;",
+      "          peg$currPos += (peg$consts[bc[ip + 1]] as string).length;",
       "          ip += 2;",
       "          break;",
       "",
       "        case " + op.FAIL + ":", // FAIL e
       "          stack.push(peg$FAILED);",
       "          if (peg$silentFails === 0) {",
-      "            peg$fail(peg$consts[bc[ip + 1]]);",
+      "            peg$fail(peg$consts[bc[ip + 1]] as ILiteralExpectation);",
       "          }",
       "          ip += 2;",
       "          break;",
@@ -1018,7 +1018,7 @@ function generateTS(ast, options) {
 
       parts.push([
         "  const peg$startRuleIndices = " + startRuleIndices + ";",
-        "  const peg$startRuleIndex = " + startRuleIndex + ";"
+        "  let peg$startRuleIndex = " + startRuleIndex + ";"
       ].join("\n"));
     } else {
       let startRuleFunctions = "{ " +

